services/src/pro/permissions/mod.rs

```diff
@@ -150,27 +151,66 @@ pub struct PermissionListing {
     pub permission: Permission,
 }
 
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(crate)), context(suffix(PermissionDbError)))]
+pub enum PermissionDbError {
+    #[snafu(display("Permission {permission:?} for resource {resource_id:?} denied."))]
+    PermissionDenied {
+        resource_id: ResourceId,
+        permission: Permission,
+    },
+    #[snafu(display("Must be admin to perform this action."))]
+    MustBeAdmin,
+    #[snafu(display(
+        "Permission {permission:?} for resource {resource_id:?} and roles {} not found.", role_ids.iter().map(std::string::ToString::to_string).collect::<Vec<String>>().join(", ")
+    ))]
```

Comment:
You must not use the `Debug` output for `Display`.
It is important that we don't leak debug messages to the user.
You should use the `Display` output without `?` instead.

Reply:
787e9011a39a9c60ca85bc897d6cca57b5837e6f